File: examples/latest/navigate/android/Java/HERESDKUnits/heresdk-units-popupmenu/src/main/java/com/here/sdk/units/popupmenu/PopupMenuView.java
/*
 * Copyright (C) 2019-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.sdk.units.popupmenu;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.LinearLayout;

import com.here.sdk.units.core.views.UnitButton;

// The view hosting this HERE SDK Unit.
// The view consists only of a button that opens a PopupMenu.
// The unit is created after inflation and provides additional logic.
public class PopupMenuView extends LinearLayout {

    public PopupMenuUnit popupMenuUnit;

    public PopupMenuView(Context context) {
        super(context);
        init(context);
    }

    public PopupMenuView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public PopupMenuView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {
        // Create button programmatically.
        UnitButton button = new UnitButton(context);

        // Add button to this view.
        addView(button);

        popupMenuUnit = new PopupMenuUnit(button, context);
    }
}

File: examples/latest/navigate/android/Kotlin/HERESDKUnitsKotlin/here-sdk-units-popupmenu/src/main/java/com/here/sdk/units/popupmenu/PopupMenuView.kt
/*
 * Copyright (C) 2019-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.sdk.units.popupmenu

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import com.here.sdk.units.core.views.UnitButton

// The view hosting this HERE SDK Unit.
// The view consists only of a button that opens a PopupMenu.
// The unit is created after inflation and provides additional logic.
@Composable
fun PopupMenuView(
    buttonText: String,
    menuItems: Map<String, () -> Unit>,
    modifier: Modifier = Modifier
) {
    var expanded by remember { mutableStateOf(false) }

    Box(
        modifier = modifier,
        contentAlignment = Alignment.Center
    ) {
        UnitButton(
            text = buttonText,
            onClick = { expanded = true },
            modifier = modifier
        )

        DropdownMenu(
            expanded = expanded,
            onDismissRequest = { expanded = false }
        ) {
            menuItems.forEach { (title, action) ->
                DropdownMenuItem(
                    text = {
                        Text(
                            text = title,
                            modifier = Modifier.fillMaxWidth(),
                            textAlign = TextAlign.Center
                        )
                    },
                    onClick = {
                        expanded = false
                        action()
                    }
                )
            }
        }
    }
}

